### PR TITLE
#66 distribute the read operation on Redis Cache

### PR DESCRIPTION
**Pull request by ryusoft**
For supporting Read Replica in Cache usage

| Q             | A
| ------------- | ---
| Is bugfix?    | no
| New feature?  | yes
| Breaks BC?    | no
| Tests pass?   | yes/no
| Fixed issues  | #66 multiple server support

### CODE REVIEW

File: Cache.php [hunk 1]
```diff
@@ -195,4 +231,47 @@ protected function flushValues()
...
+        if (count($this->replicas) >= 2) {
+            shuffle($this->replicas);
+        }
+        $config = array_shift($this->replicas);
+        if (is_array($config)) {
+            if (!isset($config['class'])) {
+                $config['class'] = 'yii\redis\Connection';
+            }
+
+            //--- if hostname same, no need re-open connection
```
Comment (samdark) [inline Cache.php]: // If hostname is the same, there is no need re-open connection

**Comment by cebe** [comment]
I have simplified the code a bit and added more docs. `Instance::ensure()` implements most of what was implemented here, so we can just use it to create the instance. Skipping the connection when the host is the same does not make sense as there can be different ports and db settings.

**Comment by cebe** [comment]
seems I broke the test... :-/

File: tests/RedisCacheTest.php [hunk 2]
```diff
@@ -120,4 +126,73 @@ public function testMultiByteGetAndSet()
...
+        $cache->set($key, $value);
+        $this->assertSame($cache->get($key), $value);
+
+        //invalid config
+        $this->resetCacheInstance();
+        $cache = $this->getCacheInstance();
+        $cache->enableReplicas = true;
+
+        $cache->replicas = [
+            ['class' => 'yii\db\Connection'],
```
Comment (cebe) [inline tests/RedisCacheTest.php]: this should expect an exception.

Comment (ryusoft) [reply tests/RedisCacheTest.php]: My initial intention of not using Instance::ensure() is not to break the application if replica is not properly configured, will update the test to expect exception

**Comment by cebe** [comment]
Thank you!

**Comment by blind3dd** [comment]
nice!.

**Pull request merged**